oap-server/server-storage-plugin/storage-elasticsearch-plugin/src/main/java/org/apache/skywalking/oap/server/storage/plugin/elasticsearch/base/StorageEsInstaller.java

```diff
@@ -68,40 +78,87 @@ protected boolean isExists(Model model) throws StorageException {
     @Override
     protected void createTable(Model model) throws StorageException {
         ElasticSearchClient esClient = (ElasticSearchClient) client;
-
         Map<String, Object> settings = createSetting(model);
         Map<String, Object> mapping = createMapping(model);
-        log.info("index {}'s columnTypeEsMapping builder str: {}", esClient.formatIndexName(model.getName()), mapping
-            .toString());
+        String tableName = PhysicalIndexManager.INSTANCE.getTableName(model);
+        PhysicalIndices.registerRelation(model.getName(), tableName);
+        log.info("index {}'s columnTypeEsMapping builder str: {}",
+                 esClient.formatIndexName(tableName), mapping.toString()
+        );
 
+        String indexName;
         try {
-            String indexName;
-            if (!model.isTimeSeries()) {
-                indexName = model.getName();
-            } else {
-                if (!esClient.isExistsTemplate(model.getName())) {
-                    boolean isAcknowledged = esClient.createTemplate(model.getName(), settings, mapping);
-                    log.info(
-                        "create {} index template finished, isAcknowledged: {}", model.getName(), isAcknowledged);
+            if (model.isTimeSeries()) {
+                if (!esClient.isExistsTemplate(tableName) || !isTemplateMappingCompatible(tableName, mapping)) {
+                    Map<String, Object> templateMapping = appendTemplateMapping(tableName, mapping);
+                    boolean isAcknowledged = esClient.putTemplate(tableName, settings, templateMapping);
+                    log.info("create {} index template finished, isAcknowledged: {}", tableName, isAcknowledged);
                     if (!isAcknowledged) {
-                        throw new StorageException("create " + model.getName() + " index template failure, ");
+                        throw new StorageException("create " + tableName + " index template failure, ");
                     }
                 }
                 indexName = TimeSeriesUtils.latestWriteIndexName(model);
+            } else {
+                indexName = tableName;
             }
-            if (!esClient.isExistsIndex(indexName)) {
-                boolean isAcknowledged = esClient.createIndex(indexName);
+
+            if (esClient.isExistsIndex(indexName)) {
```

Comment:
If not exist, don't create a new index?

Reply:
fixed

Reply:
Still not correct, I think. If you are going to delete an existing index, all data will be deleted in every rebooting process. I think you have 2 options
1. Don't recreate index, like we used to. Make the deleting as the user's responsibility
2. Deleting index only when columns don't match as expected like template. But still, you could make the OAP accidentally removes all existing data, when users are not aware.

So, I recommend (1). What do you think?

Reply:
i add a tag template called install_sw_template_finished to avoid this condition.

Reply:
I am confused about you are saying. I mean you are deleting the index every time no matter it is old version or new version.

Reply:
The index you created is an empty index, all column related things are in the template(such as mapping), am I right? 
Then why do we need to keep deleting/recreating? Even more for the indices not time series?

Reply:
The whole `INSTALL_TEMPLATE_FINISHED_TAG` thing should be removed/reverted, from my understanding. It is quite confusing. I can't find a reason to delete an index from OAP perspective, except for TTL case.

Reply:
We need use ESInstaller to check whether this index was created through the current booting process,  if yes, it is safe to delete and recreate. Otherwise, please terminate the booting with an error.
The compatibility check should be on index and template. If existing template and index have the required columns and mapping, then  booting could continue. Extra columns are not the isssue.
We could set a checker, in the storage completed stage, to check whether all physical indices matching the column. Only check, don't manipulate.

Reply:
Read https://github.com/apache/skywalking/pull/6499#discussion_r588958394 for code structure recommendation.